staging/src/k8s.io/dynamic-resource-allocation/resourceslice/resourceslicecontroller.go

```diff
@@ -381,66 +431,183 @@ func (c *Controller) syncPool(ctx context.Context, poolName string) error {
+			if _, err := c.kubeClient.ResourceV1alpha3().ResourceSlices().Update(ctx, slice, metav1.UpdateOptions{}); err != nil {
+				return fmt.Errorf("update resource slice: %w", err)
+			}
+		}
+
+		// Create new slices.
+		for i := 0; i < len(pool.Slices); i++ {
+			if _, ok := currentSliceForDesiredSlice[i]; ok {
+				// Was handled above through an update.
+				continue
+			}
+			desiredSlice := &pool.Slices[i]
+
+			var ownerReferences []metav1.OwnerReference
+			if c.owner != nil {
+				ownerReferences = append(ownerReferences,
+					metav1.OwnerReference{
+						APIVersion: c.owner.APIVersion,
+						Kind:       c.owner.Kind,
+						Name:       c.owner.Name,
+						UID:        c.owner.UID,
+						Controller: ptr.To(true),
 					},
+				)
+			}
+			generateName := c.driverName + "-"
+			if c.owner != nil {
+				generateName = c.owner.Name + "-" + generateName
+			}
+			slice := &resourceapi.ResourceSlice{
+				ObjectMeta: metav1.ObjectMeta{
+					OwnerReferences: ownerReferences,
+					GenerateName:    generateName,
+				},
+				Spec: resourceapi.ResourceSliceSpec{
+					Driver:       c.driverName,
+					Pool:         desiredPool,
+					NodeName:     nodeName,
+					NodeSelector: pool.NodeSelector,
+					AllNodes:     desiredAllNodes,
+					Devices:      desiredSlice.Devices,
 				},
 			}
-		}
 
-		slice := slices[0]
-		slice.OwnerReferences = []metav1.OwnerReference{{
-			APIVersion: c.owner.APIVersion,
-			Kind:       c.owner.Kind,
-			Name:       c.owner.Name,
-			UID:        c.owner.UID,
-			Controller: ptr.To(true),
-		}}
-		slice.Spec.Driver = c.driver
-		slice.Spec.Pool.Name = poolName
-		slice.Spec.Pool.Generation = generation
-		slice.Spec.Pool.ResourceSliceCount = 1
-		slice.Spec.NodeName = nodeName
-		slice.Spec.NodeSelector = pool.NodeSelector
-		slice.Spec.AllNodes = pool.NodeSelector == nil && nodeName == ""
-		slice.Spec.Devices = pool.Devices
-
-		if loggerV := logger.V(6); loggerV.Enabled() {
-			// Dump entire resource information.
-			loggerV.Info("Syncing resource slices", "obsoleteSlices", klog.KObjSlice(obsoleteSlices), "slices", klog.KObjSlice(slices), "pool", pool)
-		} else {
-			logger.V(5).Info("Syncing resource slices", "obsoleteSlices", klog.KObjSlice(obsoleteSlices), "slices", klog.KObjSlice(slices), "numDevices", len(pool.Devices))
+			// It can happen that we create a missing slice, some
+			// other change than the create causes another sync of
+			// the pool, and then a second slice for the same set
+			// of devices gets created because the controller has
+			// no copy of the first slice instance in its informer
+			// cache yet.
+			//
+			// This is not a problem: a client will either see one
+			// of the two slices and use it or see both and do
+			// nothing because of the duplicated device IDs.
+			//
+			// To avoid creating a second slice, we would have to use a
+			// https://pkg.go.dev/k8s.io/client-go/tools/cache#MutationCache.
```

Comment:
One unit test failed because of this race:
```
    resourceslicecontroller.go:526: I1025 22:13:35.488207] Bumped generation by one poolName="pool" generation=2
    resourceslicecontroller.go:544: I1025 22:13:35.488351] Updating existing resource slice poolName="pool" slice="resource-slice-2"
    resourceslicecontroller.go:544: I1025 22:13:35.488591] Updating existing resource slice poolName="pool" slice="resource-slice-3"
    resourceslicecontroller.go:298: I1025 22:13:35.488812] ResourceSlice update slice="resource-slice-2"
    resourceslicecontroller.go:298: I1025 22:13:35.488936] ResourceSlice update slice="resource-slice-3"
    resourceslicecontroller.go:544: I1025 22:13:35.488992] Updating existing resource slice poolName="pool" slice="resource-slice-1"
    resourceslicecontroller.go:434: I1025 22:13:35.489234] Existing slices poolName="pool" obsolete=["resource-slice-1"] current=["resource-slice-2","resource-slice-3"]
```

All three existing slices where updated, but the informer cache had only been updated with two updated slices. The third one was the original one with `generation: 1`, and therefore it was considered obsolete when another sync got started because of the update events.

While the end result was okay, the race caused more API calls - not what we want!

So we need the MutationCache after all.

Reply:
This doesn't seem to happen every time. I tested it locally and it seems to happen in some cases at irregular intervals.

Reply:
It's a classic race condition... I had tested with `go test -count=10` locally without hitting it, but that doesn't mean that there's no race :hot_face:

Reply:
This should now be complete, including an E2E test.

Reply:
I tested it locally several times and it seems that there is no longer a race error by using `cache.NewIntegerResourceVersionMutationCache` . 😄